Code/Framework/AzFramework/AzFramework/Physics/ShapeConfiguration.cpp

```diff
@@ -413,6 +413,11 @@ namespace Physics
         return m_samples;
     }
 
+    AZStd::vector<Physics::HeightMaterialPoint>& HeightfieldShapeConfiguration::GetSamples()
+    {
+        return m_samples;
+    }
+
```

Comment:
Is it possible to make the code work without requiring a non-const version of this? If you need to modify the samples, it seems like an API to modify them would be preferable over letting callers modify the underlying data directly, since with this you're losing encapsulation.

Reply:
I didn't have a specific API in mind, it's dependent on what would work best here, though if we add a ModifySample I'd suggest using row, col as parameters instead of index. But to demonstrate the concern, right now it would be possible to do the following:
```
auto& samples = config.GetSamples();
samples.resize(newRows * newCols);
```
This would get the samples out of sync with the config's m_numRows and m_numColumns values without any error-checking. Better encapsulation would make it possible to prevent this.

Looking at the code flow, I understand _why_ it's like this though. We're trying to get PhysX ShapeConfiguration data passed directly back through to the Terrain Physics Heightfield provider for modifications without making copies of the data back and forth and without the Terrain code having direct knowledge of the PhysX HeightfieldShapeConfiguration class. Here are the options I can think of:
1. Stick with the current version, and accept the lost encapsulation. I have a bad feeling this is going to cause us problems, especially when we start multi-threading, and that it will be harder to deal with those problems. But I can't say for a _fact_ that it's a problem, it's just a feeling.
2. Accept data copies (memory penalty). With this route, I'd suggest that instead of getting samples and calling an UpdateHeightAndMaterials to update a subset, just have an API to _get_ the subset of data and then use that to update the samples, either by passing the subset into an API on HeightfieldShapeConfiguration, or using something like the proposed ModifySample() API to modify each sample individually.
3. Pass through a callback (CPU penalty). This would consist of handing a per-position callback into the UpdateHeightAndMaterials that would get the new value and call something like the ModifySample() API.

I won't block the PR over any of these choices since there are justifiable reasons for choosing any of them, I just wanted to at least raise the concern and have the conversation. My gut feeling is that option 2 or 3 will work out better for us in the long run than option 1 though.

Reply:
Yes, the example of resizing the returned vector makes sense and doesn't look safe. 
Changed it to use ModifySample now, thanks!